.changeset/quiet-suits-count.md

```diff
@@ -0,0 +1,55 @@
+---
+'@ithaka/pharos': major
+'@ithaka/pharos-cli': minor
+'@ithaka/pharos-site': minor
+---
+
+require manual component registration:
+
+* Remove component self-registration
+* Scope registries for components composed of other Pharos components using [scoped-registry-mixin](https://github.com/lit/lit/tree/main/packages/labs/scoped-registry-mixin)
+* Update React wrapper to detect defined custom tag name
+* Create `PharosElement` base class which sets a custom attribute `data-pharos-component` as a common identifier
+* Set components to extend `PharosElement`
+* Update documentation
+
+**Migration Guidelines**
+
+1. Remove all individual imports of Pharos web components.
+
+2. Register them with a custom tag in your app's entrypoint:
+
+    ```javascript
+    import { PharosAlert } from '@ithaka/pharos/lib/components/alert/pharos-alert';
+
+    customElements.define('pharos-alert-homepage', PharosAlert);
+    ```
```

Comment:
I wonder if a prefix is a better recommendation (here and in the docs), as it then acts more like a namespace. That is:

```html
<homepage-pharos-alert>
<homepage-pharos-dropdown-menu>
...
```

is easier to read/reason about than:

```html
<pharos-alert-homepage>
<pharos-dropdown-menu-homepage>
...
```

---

We could also provide a little helper (_at some point_) that could make it easier for folks who need to register a big bunch of components:

```es6
import { registerComponents } from '@ithaka/pharos/lib/utils';

import { PharosAlert } from ...
import { PharosDropdownMenu } from ...

// spits out <homepage-pharos-alert>, <homepage-pharos-dropdown-menu>, ...
registerComponents('homepage', [ PharosAlert, PharosDropdownMenu, ...])
```

The consumer could do the following, I suppose? Not sure how fun `component-name-converted-to-kebab-case` will be at first glance 🤷🏻‍♂️ 

```es6
import { PharosAlert } from ...
import { PharosDropdownMenu } from ...

[ PharosAlert, PharosDropdownMenu, ...].forEach((component) => {
    customElements.define(`homepage-${<component-name-converted-to-kebab-case>}`, component);
});
```

Reply:
No strong feeling about the convention. Curious what other folks gravitate toward.

Reply:
The prefix convention further reduces the risk of collision because, when choosing a prefix, it's easier to look through existing code to understand whether that prefix is in use. As a suffix, the scannability is lost.

Reply:
To prevent collisions for the same component + different versions, should we also add the version to the naming convention?

Example: globally registered `PharosAlert` on version 2.0.0, bundle registered `PharosAlert` on version 1.0.0.

Reply:
🤔  I guess it's not needed since a scoped custom element is essentially doing that with the shadow DOM.

Reply:
Not sure I understand. Namespacing to a bundle/app will prevent any chance of collision regardless of version (even with a collision the browser will now throw an error).

Reply:
Yup, I realized that after reading the specs for how the name-spacing working. Didn't realize the scoping would account for that.